README.md
f2fclient
=========

Working on Python REST client for Food2Fork http://food2fork.com/


CHANGES.txt
v0.1 September 16, 2013 -- Initial release.

food2forkclient/tests.py
# -*- coding: utf-8 -*-
import os
import sys
import unittest


from client import (
    Food2ForkClient, Food2ForkHTTPError,
    Food2ForkSocketError, Food2ForkClientException
)

try:
    API_KEY = os.environ['API_KEY']
except KeyError:
    API_KEY = None
    sys.stderr.write('Please set os.environ["API_KEY"] = yourapikey, '
                     'or pass api_key param in Food2ForkClient')


class TestFood2ForkClient(unittest.TestCase):

    def setUp(self):
        self.f2fclient = Food2ForkClient(api_key=API_KEY)

    def test_api_url(self):
        self.assertEqual(
            self.f2fclient.URL_API, 'http://food2fork.com/api'
        )

    def test_search_url(self):
        self.assertEqual(self.f2fclient.URL_SEARCH,
                         'http://food2fork.com/api/search/?')

    def test_get_url(self):
        self.assertEqual(self.f2fclient.URL_GET,
                         'http://food2fork.com/api/get/?')

    def test_search(self):
        response = self.f2fclient.search()
        self.assertTrue(
            len(response) == 30
        )

    def test_search_count(self):
        response = self.f2fclient.search(count=10)
        self.assertTrue(
            len(response) == 10
        )

    def test_search_query(self):
        response = self.f2fclient.search(q='chicken')
        self.assertTrue(
            len(response) > 0
        )

    def test_search_params(self):
        with self.assertRaises(Food2ForkHTTPError) as cm:
            self.f2fclient.search(page='asdfasdfas')
        e = cm.exception
        self.assertEqual(e.code, 500)

    def test_page_number(self):
        self.assertRaises(
            Food2ForkClientException,
            self.f2fclient.search,
            page=99999999999999
        )

    def test_get(self):
        response = self.f2fclient.get('26851')
        self.assertTrue(len(response) > 1)

    def test_get_error(self):
        self.assertRaises(
            Food2ForkClientException,
            self.f2fclient.get,
            'asdfasdfjl;sndhdre  qwerqwasdf332323232'
            )


class TestFood2ForkAPIKeyError(unittest.TestCase):

    def setUp(self):
        self.f2fclient = Food2ForkClient(api_key='aaaaaaaaaaaaaaaaaaaaaaaaaaa')

    def test_api_key_error(self):
        with self.assertRaises(Food2ForkHTTPError) as cm:
            self.f2fclient.search()
        e = cm.exception
        self.assertEqual(e.code, 403)


class TestFood2ForkTimeoutError(unittest.TestCase):

    def setUp(self):
        self.f2fclient = Food2ForkClient(api_key=API_KEY, timeout=0.000000001)

    def test_timeout_error(self):
        with self.assertRaises(Food2ForkSocketError) as cm:
            self.f2fclient.search()
        e = cm.exception
        self.assertTrue(isinstance(e, Food2ForkSocketError))


if __name__ == '__main__':
    unittest.main()


food2forkclient/tests/tests.py
# -*- coding: utf-8 -*-
import os
import sys
import unittest


from ..client import (
    Food2ForkClient, Food2ForkHTTPError,
    Food2ForkSocketError, Food2ForkClientException
)

try:
    API_KEY = os.environ['API_KEY']
except KeyError:
    API_KEY = None
    sys.stderr.write('Please set os.environ["API_KEY"] = yourapikey, '
                     'or pass api_key param in Food2ForkClient')


class TestFood2ForkClient(unittest.TestCase):

    def setUp(self):
        self.f2fclient = Food2ForkClient(api_key=API_KEY)

    def test_api_url(self):
        self.assertEqual(
            self.f2fclient.URL_API, 'http://food2fork.com/api'
        )

    def test_search_url(self):
        self.assertEqual(self.f2fclient.URL_SEARCH,
                         'http://food2fork.com/api/search/?')

    def test_get_url(self):
        self.assertEqual(self.f2fclient.URL_GET,
                         'http://food2fork.com/api/get/?')

    def test_search(self):
        response = self.f2fclient.search()
        self.assertTrue(
            len(response) == 30
        )

    def test_search_count(self):
        response = self.f2fclient.search(count=10)
        self.assertTrue(
            len(response) == 10
        )

    def test_search_query(self):
        response = self.f2fclient.search(q='chicken')
        self.assertTrue(
            len(response) > 0
        )

    def test_search_params(self):
        with self.assertRaises(Food2ForkHTTPError) as cm:
            self.f2fclient.search(page='asdfasdfas')
        e = cm.exception
        self.assertEqual(e.code, 500)

    def test_page_number(self):
        self.assertRaises(
            Food2ForkClientException,
            self.f2fclient.search,
            page=99999999999999
        )

    def test_get(self):
        response = self.f2fclient.get('26851')
        self.assertTrue(len(response) == 1)

    def test_get_error(self):
        self.assertRaises(
            Food2ForkClientException,
            self.f2fclient.get,
            'asdfasdfjl;sndhdre  qwerqwasdf332323232'
            )


class TestFood2ForkAPIKeyError(unittest.TestCase):

    def setUp(self):
        self.f2fclient = Food2ForkClient(api_key='aaaaaaaaaaaaaaaaaaaaaaaaaaa')

    def test_api_key_error(self):
        with self.assertRaises(Food2ForkHTTPError) as cm:
            self.f2fclient.search()
        e = cm.exception
        self.assertEqual(e.code, 403)


class TestFood2ForkTimeoutError(unittest.TestCase):

    def setUp(self):
        self.f2fclient = Food2ForkClient(api_key=API_KEY, timeout=0.000000001)

    def test_timeout_error(self):
        with self.assertRaises(Food2ForkSocketError) as cm:
            self.f2fclient.search()
        e = cm.exception
        self.assertTrue(isinstance(e, Food2ForkSocketError))


if __name__ == '__main__':
    unittest.main()


food2forkclient/client.py
# -*- coding: utf-8 -*-
import json
import os
import socket
import sys
try:
    import httplib
except ImportError:
    import http.client as httplib
try:
    from urllib import urlencode
except ImportError:
    from urllib.parse import urlencode
try:
    from urllib2 import HTTPError, URLError
except ImportError:
    from urllib.error import HTTPError, URLError
try:
    import urllib2
except ImportError:
    import urllib.request as urllib2
try:
    import urlparse
except ImportError:
    import urllib.parse as urlparse


try:
    API_KEY = os.environ['API_KEY']
except KeyError:
    API_KEY = None
    sys.stderr.write('Please set os.environ["API_KEY"] = yourapikey, '
                     'or pass api_key param in Food2ForkClient')


def error_handler(fn):
    def request_wrapper(self, *args, **kwargs):
        """
        add repeat requests for timeout
        """
        try:
            response = fn(self, *args, **kwargs)
        except HTTPError as e:
            raise Food2ForkHTTPError(e)
        except URLError as e:
            if isinstance(e.reason, socket.timeout):
                msg = '{0}'.format(e.reason)
                raise Food2ForkSocketError(msg)
            else:
                msg = 'URLError - {0}'.format(e.reason)
                raise Food2ForkClientException(msg)
        except httplib.HTTPException:
            raise Food2ForkHTTPException('HTTPException')
        except Exception:
            import traceback
            msg = 'Exception - {0}'.format(traceback.format_exc())
            raise Food2ForkClientException(msg)
        if response.code != 200:
            raise Food2ForkClientException('Problem with Food2Fork API')
        return response
    return request_wrapper


def user_error_handler(fn):
    def response_wrapper(self, response):
        python_response = fn(self, response)
        parsed_url = urlparse.urlparse(response.url)
        path = parsed_url.path
        error = python_response.get('error', '')
        if error:
            raise Food2ForkClientException('API call limit exceded')
        elif path == '/api/search/':
            results = python_response.get('recipes', '')
            if not results:
                raise Food2ForkClientException('Page # does not exist')
        elif path == '/api/get/':
            results = python_response.get('recipe', '')
            if not results:
                raise Food2ForkClientException('Recipe id does not exist')
        return results
    return response_wrapper


class Food2ForkClientException(Exception):
    pass


class Food2ForkHTTPException(Exception):
    pass


class Food2ForkHTTPError(Exception):

    def __init__(self, value):
        error = value
        if error.code == 403:
            self.code = 403
            self.value = '403 Check API key'
        elif error.code == 500:
            self.code = 500
            self.value = '500 Invalid search params?'
        else:
            self.value = '{0} {1}'.format(error.code, error.reason)

    def __str__(self):
        return repr(self.value)


class Food2ForkSocketError(Exception):

    def __init__(self, value):
        self.value = value

    def __str__(self):
        return repr(self.value)


class Food2ForkClient(object):
    URL_API = 'http://food2fork.com/api'
    URL_SEARCH = URL_API + '/search/?'
    URL_GET = URL_API + '/get/?'
    HEADERS = {"Content-Type": "application/json"}

    def __init__(self, api_key=API_KEY, timeout=None):
        self.api_key = api_key
        self.timeout = timeout
        msg = ('Must pass api_key, or set '
               'os.environ["API_KEY"] = yourapikey')
        assert(api_key is not None), msg

    def search(self, q=None, page=1, count=30):
        """
        kwargs:
        q: search_query
        page: used to get additional results
        count: number of results per search
        """
        assert(0 < count <= 30), 'max 30 results per call, min 1'
        query_params = [
            ('page', page),
            ('count', count)
        ]
        if q is not None:
            query_params.append(('q', q))
        query_params.append(('key', self.api_key))
        query_string = urlencode(query_params)
        url = self.URL_SEARCH + query_string
        response = self._request(url)
        return self._parse_json(response)

    def get(self, rid):
        """
        rid: rId (recipe_id) of recipe returned by search query
        """
        query_params = [('key', self.api_key), ('rId', rid)]
        query_string = urlencode(query_params)
        url = self.URL_GET + query_string
        response = self._request(url)
        return self._parse_json(response)

    @error_handler
    def _request(self, url):
        req = urllib2.Request(url)
        for key, value in self.HEADERS.items():
            req.add_header(key, value)
        response = urllib2.urlopen(req, timeout=self.timeout)
        return response

    @user_error_handler
    def _parse_json(self, response):
        #encoding = response.headers.get_content_charset()
        python_response = json.loads(response.read().decode('utf-8'))
        return python_response

#### 403 FORRBIDEN - bad key
#### 500 Internal Server Error page or count = None
####  {u'error': u'limit'}
